src/app/icd/ICDManager.cpp

```diff
@@ -228,5 +254,74 @@ void ICDManager::OnTransitionToIdle(System::Layer * aLayer, void * appState)
     pIcdManager->mStateObserver->OnTransitionToIdle();
 }
 
+bool ICDManager::EnsureChipStackLock()
+{
+    // TODO implement platform agnostic lock creation/mechanism that can be used for ICD without locking the chip stack
+    bool requireLock = false;
+
+#if CHIP_STACK_LOCK_TRACKING_ENABLED
+    requireLock = !DeviceLayer::PlatformMgr().IsChipStackLockedByCurrentThread();
+#endif
+
+    if (requireLock)
+    {
+        DeviceLayer::PlatformMgr().LockChipStack();
+    }
+    return requireLock;
+}
+
+/* ICDNotify observer functions. */
+void ICDManager::KeepActiveRequest(KeepActiveFlags request, bool set)
+{
+    bool requiresUnlock = this->EnsureChipStackLock();
+
+    if (request == KeepActiveFlags::kExchangeContextOpen)
```

Comment:
nesting here feels too deep because of lack of lock RAII (or lock being needed in general). After RAII/nolock switch, please make this early-return as needed.